pkg/datasource.go

```diff
@@ -153,29 +161,66 @@ func (o *OCIDatasource) QueryData(ctx context.Context, req *backend.QueryDataReq
 }
 
 func (o *OCIDatasource) testResponse(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
-	return &backend.QueryDataResponse{}, nil
-	// var ts GrafanaCommonRequest
-
-	// query := req.Queries[0]
-	// if err := json.Unmarshal(query.JSON, &ts); err != nil {
-	// 	return &backend.QueryDataResponse{}, err
-	// }
-
-	// //o.logger.Error("ts.Com is " + ts.Compartment)
-	// listMetrics := monitoring.ListMetricsRequest{
-	// 	CompartmentId: common.String(ts.Compartment),
-	// }
-	// reg := common.StringToRegion(ts.Region)
-	// o.metricsClient.SetRegion(string(reg))
-	// res, err := o.metricsClient.ListMetrics(ctx, listMetrics)
-	// if err != nil {
-	// 	return &backend.QueryDataResponse{}, err
-	// }
-	// status := res.RawResponse.StatusCode
-	// if status >= 200 && status < 300 {
-	// 	return &backend.QueryDataResponse{}, nil
-	// }
-	// return nil, errors.Wrap(err, fmt.Sprintf("list metrircs failed %s %d", spew.Sdump(res), status))
+
+	// GrafanaCommonRequest - captures the common parts of the search and metricsRequests
+	// the structure is used here to retrieve the TenancyOCID and CompartmentOCID
+	var ts GrafanaCommonRequest
+	query0 := req.Queries[0]
+
+	if err := json.Unmarshal(query0.JSON, &ts); err != nil {
+		return &backend.QueryDataResponse{}, err
+	}
+	o.logger.Debug("Testing OCI logs datasource", "TenancyOCID", ts.TenancyOCID)
+	o.logger.Debug("Testing OCI logs datasource", "CompartmentOCID", ts.Compartment)
+
+	listLogsGroup := logging.ListLogGroupsRequest{
+		CompartmentId: common.String(ts.Compartment),
+	}
+	listLogsGroups, err := o.loggingMgmtClient.ListLogGroups(ctx, listLogsGroup)
+	if err == nil {
+		for _, loggroupitem := range listLogsGroups.Items {
+			if &loggroupitem.Id != nil {
+				listLog := logging.ListLogsRequest{
+					LogGroupId: common.String(*loggroupitem.Id),
+				}
+				listLogs, err := o.loggingMgmtClient.ListLogs(ctx, listLog)
+				if err == nil {
+					for _, logitem := range listLogs.Items {
+						if &logitem.Id != nil {
+							query := `search "` + ts.Compartment + `/` + *loggroupitem.Id + `/` + *logitem.Id + `"`
+							t := time.Now()
+							t2 := t.Add(-time.Minute * 30)
+							start, _ := time.Parse(time.RFC3339, t2.Format(time.RFC3339))
+							end, _ := time.Parse(time.RFC3339, t.Format(time.RFC3339))
+							request := loggingsearch.SearchLogsRequest{SearchLogsDetails: loggingsearch.SearchLogsDetails{SearchQuery: common.String(query),
+								TimeStart:         &common.SDKTime{Time: start},
+								TimeEnd:           &common.SDKTime{Time: end},
+								IsReturnFieldInfo: common.Bool(false)},
+								Limit: common.Int(10)}
+							res, err := o.loggingSearchClient.SearchLogs(ctx, request)
+							if err == nil {
+								status := res.RawResponse.StatusCode
+								o.logger.Debug("status :", "status", status)
+								if status >= 200 && status < 300 {
+									return &backend.QueryDataResponse{}, nil
```

Comment:
While this appears to be the appropriate return value to provide for the test response, I'm not seeing any result for the datasource test for either success or failure cases. And in looking at some other data sources it would appear they do something similar. This is something we need to resolve this issue to be able to claim that we are providing a reasonable datasource test result to the user.

Reply:
I tried a number of changes to the code for both success and failure cases but didn't have any luck figuring out what was going on.

Reply:
One line in the JS code that I'm a bit suspicious of line 285 of the `src/datasource.js` file:
`}, 10).then((res) => toDataQueryResponse(res, request));`
where I think this should just be:
`}, 10).then((res) => toDataQueryResponse(res, request));`
since I don't think the format of the requests generated by the plugin frontend align with the type of data is expected to be provided with the 2nd parameter of the `toDataQueryResponse()` function.

I don't know whether this line really is a problem or it is a different problem that is not effecting the test datasource flow.

Reply:
Sorry update the comment to remove the request parameter from the `toDataQueryResponse()` function

Reply:
With that change I got the following error: `logger=plugin.oci-logs-datasource t=2022-07-20T13:42:00.78+0000 lvl=dbug msg="Testing OCI logs datasource" TenancyOCID=<hidden>
logger=plugin.oci-logs-datasource t=2022-07-20T13:42:00.78+0000 lvl=dbug msg="Testing OCI logs datasource" CompartmentOCID=
logger=plugin.oci-logs-datasource t=2022-07-20T13:42:00.78+0000 lvl=eror msg="Error during listLogsGroups" error="user agent can not be blank"
logger=context t=2022-07-20T13:42:00.78+0000 lvl=eror msg="Query data error" error="failed to query data: Failed to query data: rpc error: code = Unknown desc = user agent can not be blank" remote_addr=10.0.30.157
`